### ISSUE TITLE
Edited textures and game crashing?

### ISSUE THREAD

**Issue by carpedee** [DerPopo/UABE#157]
Hello, I have a question which may not be related to the software.

Basically I used UABE 2.1d (+ the last classdata update) to extract assets and edit textures from a game made with Unity 5.5.3p1.

I have 4 files very similar (4 different outfits for a character): I can replace textures and run the game normally for two of them, the two other ones making the game crash "game.exe caused an Access Violation (0xc0000005)"

I tried on two computers, same result and it's not an issue linked to the hardware like not enough RAM.
For the two files not working I tried to extract one of the original textures and replace it by the one I just extracted, even doing that make the game crash.

So my question is, is it possible that the game has some sort of protection against edited assets? It seems unlikely since I can edit two files without any issue...

EDIT: I made a few tests, here are the results:

- using UABE 2.1d + the last classdata update makes the game crash "access violation".
- using UABE 2.1/b/c/d (without the classdata update)  doesn't make the game crash: instead my character (wearing the modified outfit) is invisible and can't move. But the game isn't frozen, I can see stuff moving normally and I can pause/resume. More like the textures weren't loaded at all
- 32-bit and 64-bit versions don't make any differences 

I just saw that the Patch 5.5.3p1 for Unity was released after your last UABE update, could it be that UABE is simply not yet compatible then?

I have another unrelated question:

Let's say I want to swap two characters in a unity-made game, how would I proceed?
I know I can export meshes from the first character as raw files and import them as raw files into the second character's asset file. But the UVs would never match the textures though?

EDIT2: I managed to make the edited textures work using UnityEx. I suspect there is an issue with UABE with this very version of Unity  (5.5.3p1), maybe not with the software itself but maybe the texture plugin to import textures

**Comment by carpedee** [comment]
That would explain the access violation indeed. I'll keep you updated when I try your next release.

Regarding the mesh swap, I have been experiencing funny results and I'm not exactly sure why:

- I have the same character with two different outfits (originally in the game, not modified), resulting in two slightly different meshes. If I replace the raw mesh of character number one by the raw mesh of character number 2 (both have the same skeleton), it works more or less (the mesh is displayed correctly) but with obviously wrong UVs and messed up textures.
- In addition, if I replace the textures of number one character by the ones from character number 2, I get the same result, messed up textures. Both my characters share the same hair and arms textures, so those are the only ones showing up correctly, even though I replaced all the textures. I would have thought by swapping the meshes the UV would come with it, right? If only the hair and arms show up the second time, it would mean the UVs didn't follow the mesh and I'm not sure why...
- I tried to replace everything (always with export/import raw files): mesh, textures, materials. This time the game works but my character is invisible.

I must be missing something somewhere...

If I understand correctly:

- the raw mesh contains the mesh, bones, UVs and probably links to materials and textures, right?
- When I swap the mesh and textures from one character to another, I expect the textures to show up correctly since the new textures should be consistent with the new mesh which contains the correct UVs, right?
- Why is my character turning invisible when I swap the materials? Since the textures are basically applied on the materials when I swap everything (mesh, textures, materials), it should display everything correctly? There must be another file involved (shaders? skinnedmeshrenderer?) but I don't understand which one

**Comment by DerPopo** [comment]
Swapping the meshes and textures should work in most cases. The mesh contains bones and UVs but doesn't link to other assets. Instead, GameObjects link to them through MeshFilter and MeshRenderer components.
Materials contain references (PPtrs) to the textures. Those should be updated after swapping (or only swap the shader PPtr instead of swapping Material assets).
SkinnedMeshRenderers might need the m_StaticBatchInfo updated for the mesh (start index will usually be 0, count the size of the m_SubMeshes array of the Mesh). These assetsshould not be swapped, since they contain a reference to their GameObject and to Transforms also linked to that GameObject.
The only issue might be the bone transform references in SkinnedMeshRenderers. In the worst case, you might end up having to change the GameObject's component and children chain (childrens are linked through each GameObject's Transform component for some reason, making it even more inconvenient).

**Comment by carpedee** [comment]
Thank you for these detailed explanations.
The only bit that confused me was "swap the shader PPtr of the original Material instead"
Do you mean that part?

 PPtr<Shader> m_Shader
  0 int m_FileID = 0
  0 SInt64 m_PathID = "PathID"

This "PathID" refers to a shader so I'm not so sure you were talking about that.

**Comment by DerPopo** [comment]
That's what I meant. Make sure you also set the FileID accordingly, if the Materials aren't in the same .assets file. Use View->References (with the absolute FileID as in the asset list) to find out what the FileID in one Material refers to and try to find the matching FileID for the other Material.

**Comment by carpedee** [comment]
To continue with the same example I mentioned above: same character with two different outfits resulting in two different meshes, same bones and animations used for both.
Each character is inside a different bundle/.assets file (a .bytes file to be exact)

The materials used for each one link to the same PathID for "SInt64 m_PathID", meaning they are using the same shader.
The file ID for both materials is the same, "0". Is that normal?

Here is what I got when exporting the dump files:

0 Material Base
 1 string m_Name = "Charac1_body_MAP"
 0 PPtr<Shader> m_Shader
  0 int m_FileID = 0
  0 SInt64 m_PathID = 6159332043834350306
 1 string m_ShaderKeywords = "_ALPHATEST_ON _EMISSION _NORMALMAP _SPECGLOSSMAP"
 0 unsigned int m_LightmapFlags = 1
 0 int m_CustomRenderQueue = 2450
 0 map stringTagMap

and

0 Material Base
 1 string m_Name = "Charac2_body_MAP"
 0 PPtr<Shader> m_Shader
  0 int m_FileID = 0
  0 SInt64 m_PathID = 6159332043834350306
 1 string m_ShaderKeywords = "_EMISSION _NORMALMAP _SPECGLOSSMAP"
 0 unsigned int m_LightmapFlags = 1
 0 int m_CustomRenderQueue = -1
 0 map stringTagMap

I honestly don't understand why the mesh + textures swap doesn't work, especially if both characters share the same materials and shaders...

EDIT: I just realized that all the assets read "FileID = 0", no matter the type. That is weird
